{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b6fe8b5b-21f1-4bb2-9ff0-8d31fa515388",
   "metadata": {},
   "source": [
    "# Processing an eDNA dataset to Darwin Core\n",
    "## Reading the original dataset\n",
    "### List all dataset files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "de407417-e478-4003-aaaf-972f02a1c993",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'../dataset/metadata.txt'</li><li>'../dataset/samples.xlsx'</li><li>'../dataset/seqtab.txt'</li><li>'../dataset/sequences.fasta'</li><li>'../dataset/taxonomy.txt'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item '../dataset/metadata.txt'\n",
       "\\item '../dataset/samples.xlsx'\n",
       "\\item '../dataset/seqtab.txt'\n",
       "\\item '../dataset/sequences.fasta'\n",
       "\\item '../dataset/taxonomy.txt'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. '../dataset/metadata.txt'\n",
       "2. '../dataset/samples.xlsx'\n",
       "3. '../dataset/seqtab.txt'\n",
       "4. '../dataset/sequences.fasta'\n",
       "5. '../dataset/taxonomy.txt'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"../dataset/metadata.txt\"    \"../dataset/samples.xlsx\"   \n",
       "[3] \"../dataset/seqtab.txt\"      \"../dataset/sequences.fasta\"\n",
       "[5] \"../dataset/taxonomy.txt\"   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "list.files(\"../dataset\", full.names = \"TRUE\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fce83607-62f5-4d5f-a751-9a52eab894e3",
   "metadata": {},
   "source": [
    "### Read the ASV table\n",
    "\n",
    "`../dataset/seqtab.txt` contains the ASV table, so it has one row per ASV, and the number of reads in a sample in different columns."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8792c874-40b0-42d0-95bd-18be2fa0c3b4",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>asv</th><th scope=col>EE0493</th><th scope=col>EE0495</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>asv.1</td><td>    0</td><td>   0</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>asv.2</td><td>   14</td><td>2447</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>asv.3</td><td>    0</td><td>   0</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>asv.4</td><td>    0</td><td>   0</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>asv.5</td><td>40587</td><td>1857</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>asv.6</td><td>    7</td><td>  10</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 3\n",
       "\\begin{tabular}{r|lll}\n",
       "  & asv & EE0493 & EE0495\\\\\n",
       "  & <chr> & <int> & <int>\\\\\n",
       "\\hline\n",
       "\t1 & asv.1 &     0 &    0\\\\\n",
       "\t2 & asv.2 &    14 & 2447\\\\\n",
       "\t3 & asv.3 &     0 &    0\\\\\n",
       "\t4 & asv.4 &     0 &    0\\\\\n",
       "\t5 & asv.5 & 40587 & 1857\\\\\n",
       "\t6 & asv.6 &     7 &   10\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 3\n",
       "\n",
       "| <!--/--> | asv &lt;chr&gt; | EE0493 &lt;int&gt; | EE0495 &lt;int&gt; |\n",
       "|---|---|---|---|\n",
       "| 1 | asv.1 |     0 |    0 |\n",
       "| 2 | asv.2 |    14 | 2447 |\n",
       "| 3 | asv.3 |     0 |    0 |\n",
       "| 4 | asv.4 |     0 |    0 |\n",
       "| 5 | asv.5 | 40587 | 1857 |\n",
       "| 6 | asv.6 |     7 |   10 |\n",
       "\n"
      ],
      "text/plain": [
       "  asv   EE0493 EE0495\n",
       "1 asv.1     0     0  \n",
       "2 asv.2    14  2447  \n",
       "3 asv.3     0     0  \n",
       "4 asv.4     0     0  \n",
       "5 asv.5 40587  1857  \n",
       "6 asv.6     7    10  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "library(dplyr)\n",
    "\n",
    "seqtab <- read.table(\"../dataset/seqtab.txt\", sep = \"\\t\", header = TRUE)\n",
    "head(seqtab)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7eb6f801-bd0e-48b0-bcf9-8f91b4d1cfba",
   "metadata": {},
   "source": [
    "### Read the taxonomy file"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74d54e63-082d-4565-8324-953963480610",
   "metadata": {},
   "source": [
    "`../dataset/taxonomy.txt` contains a taxon name for each ASV."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8e03544a-a131-4cf2-a7c4-052374e37b6b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>asv</th><th scope=col>taxonomy</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>asv.1</td><td>Eukaryota             </td></tr>\n",
       "\t<tr><th scope=row>2</th><td>asv.2</td><td>Clausocalanus_furcatus</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>asv.3</td><td>Eurotatoria           </td></tr>\n",
       "\t<tr><th scope=row>4</th><td>asv.4</td><td>Arthropoda            </td></tr>\n",
       "\t<tr><th scope=row>5</th><td>asv.5</td><td>Eukaryota             </td></tr>\n",
       "\t<tr><th scope=row>6</th><td>asv.6</td><td>Farranula_gibbula     </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       "  & asv & taxonomy\\\\\n",
       "  & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & asv.1 & Eukaryota             \\\\\n",
       "\t2 & asv.2 & Clausocalanus\\_furcatus\\\\\n",
       "\t3 & asv.3 & Eurotatoria           \\\\\n",
       "\t4 & asv.4 & Arthropoda            \\\\\n",
       "\t5 & asv.5 & Eukaryota             \\\\\n",
       "\t6 & asv.6 & Farranula\\_gibbula     \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 2\n",
       "\n",
       "| <!--/--> | asv &lt;chr&gt; | taxonomy &lt;chr&gt; |\n",
       "|---|---|---|\n",
       "| 1 | asv.1 | Eukaryota              |\n",
       "| 2 | asv.2 | Clausocalanus_furcatus |\n",
       "| 3 | asv.3 | Eurotatoria            |\n",
       "| 4 | asv.4 | Arthropoda             |\n",
       "| 5 | asv.5 | Eukaryota              |\n",
       "| 6 | asv.6 | Farranula_gibbula      |\n",
       "\n"
      ],
      "text/plain": [
       "  asv   taxonomy              \n",
       "1 asv.1 Eukaryota             \n",
       "2 asv.2 Clausocalanus_furcatus\n",
       "3 asv.3 Eurotatoria           \n",
       "4 asv.4 Arthropoda            \n",
       "5 asv.5 Eukaryota             \n",
       "6 asv.6 Farranula_gibbula     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "taxonomy <- read.table(\"../dataset/taxonomy.txt\", sep = \"\\t\", header = TRUE)\n",
    "head(taxonomy)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9176270f-3c59-4506-b8e3-a0cdb0dca1e9",
   "metadata": {},
   "source": [
    "These names originate from the reference database and will have to be matched to WoRMS later."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1cce03aa-38eb-432c-b469-528bd98cf111",
   "metadata": {},
   "source": [
    "### Read the sample metadata\n",
    "\n",
    "We also have an Excel file with sample info."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d8360180-9da8-4fae-a085-ca37d4a152db",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 2 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>name</th><th scope=col>size</th><th scope=col>event_begin</th><th scope=col>area_name</th><th scope=col>area_longitude</th><th scope=col>area_latitude</th><th scope=col>area_uncertainty</th><th scope=col>parent_area_name</th><th scope=col>dna</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>EE0493</td><td>1450</td><td>24/04/2023</td><td>Ile esprit      </td><td>46.22536</td><td>9.425180</td><td>20</td><td>Aldabra Atoll</td><td> 7.23</td></tr>\n",
       "\t<tr><td>EE0495</td><td>1500</td><td>02/04/2023</td><td>Settlement beach</td><td>46.20605</td><td>9.400901</td><td>98</td><td>Aldabra Atoll</td><td>15.83</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 2 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " name & size & event\\_begin & area\\_name & area\\_longitude & area\\_latitude & area\\_uncertainty & parent\\_area\\_name & dna\\\\\n",
       " <chr> & <dbl> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\t EE0493 & 1450 & 24/04/2023 & Ile esprit       & 46.22536 & 9.425180 & 20 & Aldabra Atoll &  7.23\\\\\n",
       "\t EE0495 & 1500 & 02/04/2023 & Settlement beach & 46.20605 & 9.400901 & 98 & Aldabra Atoll & 15.83\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 2 × 9\n",
       "\n",
       "| name &lt;chr&gt; | size &lt;dbl&gt; | event_begin &lt;chr&gt; | area_name &lt;chr&gt; | area_longitude &lt;dbl&gt; | area_latitude &lt;dbl&gt; | area_uncertainty &lt;dbl&gt; | parent_area_name &lt;chr&gt; | dna &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| EE0493 | 1450 | 24/04/2023 | Ile esprit       | 46.22536 | 9.425180 | 20 | Aldabra Atoll |  7.23 |\n",
       "| EE0495 | 1500 | 02/04/2023 | Settlement beach | 46.20605 | 9.400901 | 98 | Aldabra Atoll | 15.83 |\n",
       "\n"
      ],
      "text/plain": [
       "  name   size event_begin area_name        area_longitude area_latitude\n",
       "1 EE0493 1450 24/04/2023  Ile esprit       46.22536       9.425180     \n",
       "2 EE0495 1500 02/04/2023  Settlement beach 46.20605       9.400901     \n",
       "  area_uncertainty parent_area_name dna  \n",
       "1 20               Aldabra Atoll     7.23\n",
       "2 98               Aldabra Atoll    15.83"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "samples <- readxl::read_excel(\"../dataset/samples.xlsx\")\n",
    "samples"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61e0e0ca-34fa-4036-b88b-a5601c0fe2f5",
   "metadata": {},
   "source": [
    "## Joining the tables\n",
    "\n",
    "At this point we could start quality control on the individual tables, but if we first join and map the tables to Darwin Core occurrence terms, the quality control code will be easier to read.\n",
    "\n",
    "### Event fields\n",
    "\n",
    "Let's start with the sample table. This table has sample identifiers, time, coordinates, coordinate uncertainty, locality, and higher geography which can all be mapped to Darwin Core."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "eb781051-e0ac-4e57-b5bd-d7dc67c062bb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 2 × 10</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>eventID</th><th scope=col>materialSampleID</th><th scope=col>eventDate</th><th scope=col>locality</th><th scope=col>decimalLongitude</th><th scope=col>decimalLatitude</th><th scope=col>coordinateUncertaintyInMeters</th><th scope=col>higherGeography</th><th scope=col>sampleSizeValue</th><th scope=col>sampleSizeUnit</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>EE0493</td><td>EE0493</td><td>24/04/2023</td><td>Ile esprit      </td><td>46.22536</td><td>9.425180</td><td>20</td><td>Aldabra Atoll</td><td>1450</td><td>ml</td></tr>\n",
       "\t<tr><td>EE0495</td><td>EE0495</td><td>02/04/2023</td><td>Settlement beach</td><td>46.20605</td><td>9.400901</td><td>98</td><td>Aldabra Atoll</td><td>1500</td><td>ml</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 2 × 10\n",
       "\\begin{tabular}{llllllllll}\n",
       " eventID & materialSampleID & eventDate & locality & decimalLongitude & decimalLatitude & coordinateUncertaintyInMeters & higherGeography & sampleSizeValue & sampleSizeUnit\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <chr> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\t EE0493 & EE0493 & 24/04/2023 & Ile esprit       & 46.22536 & 9.425180 & 20 & Aldabra Atoll & 1450 & ml\\\\\n",
       "\t EE0495 & EE0495 & 02/04/2023 & Settlement beach & 46.20605 & 9.400901 & 98 & Aldabra Atoll & 1500 & ml\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 2 × 10\n",
       "\n",
       "| eventID &lt;chr&gt; | materialSampleID &lt;chr&gt; | eventDate &lt;chr&gt; | locality &lt;chr&gt; | decimalLongitude &lt;dbl&gt; | decimalLatitude &lt;dbl&gt; | coordinateUncertaintyInMeters &lt;dbl&gt; | higherGeography &lt;chr&gt; | sampleSizeValue &lt;dbl&gt; | sampleSizeUnit &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|\n",
       "| EE0493 | EE0493 | 24/04/2023 | Ile esprit       | 46.22536 | 9.425180 | 20 | Aldabra Atoll | 1450 | ml |\n",
       "| EE0495 | EE0495 | 02/04/2023 | Settlement beach | 46.20605 | 9.400901 | 98 | Aldabra Atoll | 1500 | ml |\n",
       "\n"
      ],
      "text/plain": [
       "  eventID materialSampleID eventDate  locality         decimalLongitude\n",
       "1 EE0493  EE0493           24/04/2023 Ile esprit       46.22536        \n",
       "2 EE0495  EE0495           02/04/2023 Settlement beach 46.20605        \n",
       "  decimalLatitude coordinateUncertaintyInMeters higherGeography sampleSizeValue\n",
       "1 9.425180        20                            Aldabra Atoll   1450           \n",
       "2 9.400901        98                            Aldabra Atoll   1500           \n",
       "  sampleSizeUnit\n",
       "1 ml            \n",
       "2 ml            "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "event <- samples %>%\n",
    "    select(\n",
    "        eventID = name,\n",
    "        materialSampleID = name,\n",
    "        eventDate = event_begin,\n",
    "        locality = area_name,\n",
    "        decimalLongitude = area_longitude,\n",
    "        decimalLatitude = area_latitude,\n",
    "        coordinateUncertaintyInMeters = area_uncertainty,\n",
    "        higherGeography = parent_area_name,\n",
    "        sampleSizeValue = size\n",
    "    ) %>%\n",
    "    mutate(sampleSizeUnit = \"ml\")\n",
    "event"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "691970eb-325f-4f79-8fd2-fff54cf871a4",
   "metadata": {},
   "source": [
    "### Occurrence fields\n",
    "\n",
    "Next is the ASV table. This table is in a wide format with ASVs as rows and samples as columns. We will convert this to a long format, with one row per occurrence and the number of sequence reads as `organismQuantity`. We will use the sample identifier as `eventID` and the combination of sample identifier and ASV number as the `occurrenceID`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "ed9cb036-2cf0-422d-a983-4597cb51c06e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 13375 × 4</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>asv</th><th scope=col>eventID</th><th scope=col>organismQuantity</th><th scope=col>occurrenceID</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>asv.2 </td><td>EE0493</td><td>   14</td><td>EE0493_asv.2 </td></tr>\n",
       "\t<tr><td>asv.5 </td><td>EE0493</td><td>40587</td><td>EE0493_asv.5 </td></tr>\n",
       "\t<tr><td>asv.6 </td><td>EE0493</td><td>    7</td><td>EE0493_asv.6 </td></tr>\n",
       "\t<tr><td>asv.7 </td><td>EE0493</td><td>29367</td><td>EE0493_asv.7 </td></tr>\n",
       "\t<tr><td>asv.8 </td><td>EE0493</td><td>72378</td><td>EE0493_asv.8 </td></tr>\n",
       "\t<tr><td>asv.9 </td><td>EE0493</td><td>35970</td><td>EE0493_asv.9 </td></tr>\n",
       "\t<tr><td>asv.10</td><td>EE0493</td><td>   10</td><td>EE0493_asv.10</td></tr>\n",
       "\t<tr><td>asv.11</td><td>EE0493</td><td> 4375</td><td>EE0493_asv.11</td></tr>\n",
       "\t<tr><td>asv.12</td><td>EE0493</td><td> 6853</td><td>EE0493_asv.12</td></tr>\n",
       "\t<tr><td>asv.14</td><td>EE0493</td><td> 2299</td><td>EE0493_asv.14</td></tr>\n",
       "\t<tr><td>asv.16</td><td>EE0493</td><td>   18</td><td>EE0493_asv.16</td></tr>\n",
       "\t<tr><td>asv.17</td><td>EE0493</td><td> 8017</td><td>EE0493_asv.17</td></tr>\n",
       "\t<tr><td>asv.19</td><td>EE0493</td><td> 1736</td><td>EE0493_asv.19</td></tr>\n",
       "\t<tr><td>asv.20</td><td>EE0493</td><td>  428</td><td>EE0493_asv.20</td></tr>\n",
       "\t<tr><td>asv.21</td><td>EE0493</td><td> 1711</td><td>EE0493_asv.21</td></tr>\n",
       "\t<tr><td>asv.25</td><td>EE0493</td><td> 2539</td><td>EE0493_asv.25</td></tr>\n",
       "\t<tr><td>asv.28</td><td>EE0493</td><td>    8</td><td>EE0493_asv.28</td></tr>\n",
       "\t<tr><td>asv.29</td><td>EE0493</td><td> 2378</td><td>EE0493_asv.29</td></tr>\n",
       "\t<tr><td>asv.30</td><td>EE0493</td><td>    5</td><td>EE0493_asv.30</td></tr>\n",
       "\t<tr><td>asv.31</td><td>EE0493</td><td> 2645</td><td>EE0493_asv.31</td></tr>\n",
       "\t<tr><td>asv.33</td><td>EE0493</td><td> 2074</td><td>EE0493_asv.33</td></tr>\n",
       "\t<tr><td>asv.37</td><td>EE0493</td><td> 2200</td><td>EE0493_asv.37</td></tr>\n",
       "\t<tr><td>asv.38</td><td>EE0493</td><td> 1204</td><td>EE0493_asv.38</td></tr>\n",
       "\t<tr><td>asv.39</td><td>EE0493</td><td>   18</td><td>EE0493_asv.39</td></tr>\n",
       "\t<tr><td>asv.42</td><td>EE0493</td><td>    7</td><td>EE0493_asv.42</td></tr>\n",
       "\t<tr><td>asv.43</td><td>EE0493</td><td>  634</td><td>EE0493_asv.43</td></tr>\n",
       "\t<tr><td>asv.45</td><td>EE0493</td><td>   71</td><td>EE0493_asv.45</td></tr>\n",
       "\t<tr><td>asv.47</td><td>EE0493</td><td>  422</td><td>EE0493_asv.47</td></tr>\n",
       "\t<tr><td>asv.49</td><td>EE0493</td><td>    1</td><td>EE0493_asv.49</td></tr>\n",
       "\t<tr><td>asv.50</td><td>EE0493</td><td>    5</td><td>EE0493_asv.50</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>asv.16798</td><td>EE0495</td><td>11</td><td>EE0495_asv.16798</td></tr>\n",
       "\t<tr><td>asv.16813</td><td>EE0495</td><td> 1</td><td>EE0495_asv.16813</td></tr>\n",
       "\t<tr><td>asv.16814</td><td>EE0495</td><td> 1</td><td>EE0495_asv.16814</td></tr>\n",
       "\t<tr><td>asv.16815</td><td>EE0495</td><td> 4</td><td>EE0495_asv.16815</td></tr>\n",
       "\t<tr><td>asv.16816</td><td>EE0495</td><td> 7</td><td>EE0495_asv.16816</td></tr>\n",
       "\t<tr><td>asv.16823</td><td>EE0495</td><td> 3</td><td>EE0495_asv.16823</td></tr>\n",
       "\t<tr><td>asv.16826</td><td>EE0495</td><td> 9</td><td>EE0495_asv.16826</td></tr>\n",
       "\t<tr><td>asv.16847</td><td>EE0495</td><td> 3</td><td>EE0495_asv.16847</td></tr>\n",
       "\t<tr><td>asv.16864</td><td>EE0495</td><td> 1</td><td>EE0495_asv.16864</td></tr>\n",
       "\t<tr><td>asv.16867</td><td>EE0495</td><td> 5</td><td>EE0495_asv.16867</td></tr>\n",
       "\t<tr><td>asv.16875</td><td>EE0495</td><td> 1</td><td>EE0495_asv.16875</td></tr>\n",
       "\t<tr><td>asv.16879</td><td>EE0495</td><td> 2</td><td>EE0495_asv.16879</td></tr>\n",
       "\t<tr><td>asv.16880</td><td>EE0495</td><td> 5</td><td>EE0495_asv.16880</td></tr>\n",
       "\t<tr><td>asv.16881</td><td>EE0495</td><td> 5</td><td>EE0495_asv.16881</td></tr>\n",
       "\t<tr><td>asv.16882</td><td>EE0495</td><td> 3</td><td>EE0495_asv.16882</td></tr>\n",
       "\t<tr><td>asv.16897</td><td>EE0495</td><td> 2</td><td>EE0495_asv.16897</td></tr>\n",
       "\t<tr><td>asv.16898</td><td>EE0495</td><td> 1</td><td>EE0495_asv.16898</td></tr>\n",
       "\t<tr><td>asv.16914</td><td>EE0495</td><td> 2</td><td>EE0495_asv.16914</td></tr>\n",
       "\t<tr><td>asv.16916</td><td>EE0495</td><td> 2</td><td>EE0495_asv.16916</td></tr>\n",
       "\t<tr><td>asv.16918</td><td>EE0495</td><td> 3</td><td>EE0495_asv.16918</td></tr>\n",
       "\t<tr><td>asv.16919</td><td>EE0495</td><td> 3</td><td>EE0495_asv.16919</td></tr>\n",
       "\t<tr><td>asv.16920</td><td>EE0495</td><td> 2</td><td>EE0495_asv.16920</td></tr>\n",
       "\t<tr><td>asv.16921</td><td>EE0495</td><td> 3</td><td>EE0495_asv.16921</td></tr>\n",
       "\t<tr><td>asv.16922</td><td>EE0495</td><td> 3</td><td>EE0495_asv.16922</td></tr>\n",
       "\t<tr><td>asv.16923</td><td>EE0495</td><td> 2</td><td>EE0495_asv.16923</td></tr>\n",
       "\t<tr><td>asv.16949</td><td>EE0495</td><td> 1</td><td>EE0495_asv.16949</td></tr>\n",
       "\t<tr><td>asv.16958</td><td>EE0495</td><td> 1</td><td>EE0495_asv.16958</td></tr>\n",
       "\t<tr><td>asv.16961</td><td>EE0495</td><td> 1</td><td>EE0495_asv.16961</td></tr>\n",
       "\t<tr><td>asv.16962</td><td>EE0495</td><td> 1</td><td>EE0495_asv.16962</td></tr>\n",
       "\t<tr><td>asv.16963</td><td>EE0495</td><td> 2</td><td>EE0495_asv.16963</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 13375 × 4\n",
       "\\begin{tabular}{llll}\n",
       " asv & eventID & organismQuantity & occurrenceID\\\\\n",
       " <chr> & <chr> & <int> & <chr>\\\\\n",
       "\\hline\n",
       "\t asv.2  & EE0493 &    14 & EE0493\\_asv.2 \\\\\n",
       "\t asv.5  & EE0493 & 40587 & EE0493\\_asv.5 \\\\\n",
       "\t asv.6  & EE0493 &     7 & EE0493\\_asv.6 \\\\\n",
       "\t asv.7  & EE0493 & 29367 & EE0493\\_asv.7 \\\\\n",
       "\t asv.8  & EE0493 & 72378 & EE0493\\_asv.8 \\\\\n",
       "\t asv.9  & EE0493 & 35970 & EE0493\\_asv.9 \\\\\n",
       "\t asv.10 & EE0493 &    10 & EE0493\\_asv.10\\\\\n",
       "\t asv.11 & EE0493 &  4375 & EE0493\\_asv.11\\\\\n",
       "\t asv.12 & EE0493 &  6853 & EE0493\\_asv.12\\\\\n",
       "\t asv.14 & EE0493 &  2299 & EE0493\\_asv.14\\\\\n",
       "\t asv.16 & EE0493 &    18 & EE0493\\_asv.16\\\\\n",
       "\t asv.17 & EE0493 &  8017 & EE0493\\_asv.17\\\\\n",
       "\t asv.19 & EE0493 &  1736 & EE0493\\_asv.19\\\\\n",
       "\t asv.20 & EE0493 &   428 & EE0493\\_asv.20\\\\\n",
       "\t asv.21 & EE0493 &  1711 & EE0493\\_asv.21\\\\\n",
       "\t asv.25 & EE0493 &  2539 & EE0493\\_asv.25\\\\\n",
       "\t asv.28 & EE0493 &     8 & EE0493\\_asv.28\\\\\n",
       "\t asv.29 & EE0493 &  2378 & EE0493\\_asv.29\\\\\n",
       "\t asv.30 & EE0493 &     5 & EE0493\\_asv.30\\\\\n",
       "\t asv.31 & EE0493 &  2645 & EE0493\\_asv.31\\\\\n",
       "\t asv.33 & EE0493 &  2074 & EE0493\\_asv.33\\\\\n",
       "\t asv.37 & EE0493 &  2200 & EE0493\\_asv.37\\\\\n",
       "\t asv.38 & EE0493 &  1204 & EE0493\\_asv.38\\\\\n",
       "\t asv.39 & EE0493 &    18 & EE0493\\_asv.39\\\\\n",
       "\t asv.42 & EE0493 &     7 & EE0493\\_asv.42\\\\\n",
       "\t asv.43 & EE0493 &   634 & EE0493\\_asv.43\\\\\n",
       "\t asv.45 & EE0493 &    71 & EE0493\\_asv.45\\\\\n",
       "\t asv.47 & EE0493 &   422 & EE0493\\_asv.47\\\\\n",
       "\t asv.49 & EE0493 &     1 & EE0493\\_asv.49\\\\\n",
       "\t asv.50 & EE0493 &     5 & EE0493\\_asv.50\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t asv.16798 & EE0495 & 11 & EE0495\\_asv.16798\\\\\n",
       "\t asv.16813 & EE0495 &  1 & EE0495\\_asv.16813\\\\\n",
       "\t asv.16814 & EE0495 &  1 & EE0495\\_asv.16814\\\\\n",
       "\t asv.16815 & EE0495 &  4 & EE0495\\_asv.16815\\\\\n",
       "\t asv.16816 & EE0495 &  7 & EE0495\\_asv.16816\\\\\n",
       "\t asv.16823 & EE0495 &  3 & EE0495\\_asv.16823\\\\\n",
       "\t asv.16826 & EE0495 &  9 & EE0495\\_asv.16826\\\\\n",
       "\t asv.16847 & EE0495 &  3 & EE0495\\_asv.16847\\\\\n",
       "\t asv.16864 & EE0495 &  1 & EE0495\\_asv.16864\\\\\n",
       "\t asv.16867 & EE0495 &  5 & EE0495\\_asv.16867\\\\\n",
       "\t asv.16875 & EE0495 &  1 & EE0495\\_asv.16875\\\\\n",
       "\t asv.16879 & EE0495 &  2 & EE0495\\_asv.16879\\\\\n",
       "\t asv.16880 & EE0495 &  5 & EE0495\\_asv.16880\\\\\n",
       "\t asv.16881 & EE0495 &  5 & EE0495\\_asv.16881\\\\\n",
       "\t asv.16882 & EE0495 &  3 & EE0495\\_asv.16882\\\\\n",
       "\t asv.16897 & EE0495 &  2 & EE0495\\_asv.16897\\\\\n",
       "\t asv.16898 & EE0495 &  1 & EE0495\\_asv.16898\\\\\n",
       "\t asv.16914 & EE0495 &  2 & EE0495\\_asv.16914\\\\\n",
       "\t asv.16916 & EE0495 &  2 & EE0495\\_asv.16916\\\\\n",
       "\t asv.16918 & EE0495 &  3 & EE0495\\_asv.16918\\\\\n",
       "\t asv.16919 & EE0495 &  3 & EE0495\\_asv.16919\\\\\n",
       "\t asv.16920 & EE0495 &  2 & EE0495\\_asv.16920\\\\\n",
       "\t asv.16921 & EE0495 &  3 & EE0495\\_asv.16921\\\\\n",
       "\t asv.16922 & EE0495 &  3 & EE0495\\_asv.16922\\\\\n",
       "\t asv.16923 & EE0495 &  2 & EE0495\\_asv.16923\\\\\n",
       "\t asv.16949 & EE0495 &  1 & EE0495\\_asv.16949\\\\\n",
       "\t asv.16958 & EE0495 &  1 & EE0495\\_asv.16958\\\\\n",
       "\t asv.16961 & EE0495 &  1 & EE0495\\_asv.16961\\\\\n",
       "\t asv.16962 & EE0495 &  1 & EE0495\\_asv.16962\\\\\n",
       "\t asv.16963 & EE0495 &  2 & EE0495\\_asv.16963\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 13375 × 4\n",
       "\n",
       "| asv &lt;chr&gt; | eventID &lt;chr&gt; | organismQuantity &lt;int&gt; | occurrenceID &lt;chr&gt; |\n",
       "|---|---|---|---|\n",
       "| asv.2  | EE0493 |    14 | EE0493_asv.2  |\n",
       "| asv.5  | EE0493 | 40587 | EE0493_asv.5  |\n",
       "| asv.6  | EE0493 |     7 | EE0493_asv.6  |\n",
       "| asv.7  | EE0493 | 29367 | EE0493_asv.7  |\n",
       "| asv.8  | EE0493 | 72378 | EE0493_asv.8  |\n",
       "| asv.9  | EE0493 | 35970 | EE0493_asv.9  |\n",
       "| asv.10 | EE0493 |    10 | EE0493_asv.10 |\n",
       "| asv.11 | EE0493 |  4375 | EE0493_asv.11 |\n",
       "| asv.12 | EE0493 |  6853 | EE0493_asv.12 |\n",
       "| asv.14 | EE0493 |  2299 | EE0493_asv.14 |\n",
       "| asv.16 | EE0493 |    18 | EE0493_asv.16 |\n",
       "| asv.17 | EE0493 |  8017 | EE0493_asv.17 |\n",
       "| asv.19 | EE0493 |  1736 | EE0493_asv.19 |\n",
       "| asv.20 | EE0493 |   428 | EE0493_asv.20 |\n",
       "| asv.21 | EE0493 |  1711 | EE0493_asv.21 |\n",
       "| asv.25 | EE0493 |  2539 | EE0493_asv.25 |\n",
       "| asv.28 | EE0493 |     8 | EE0493_asv.28 |\n",
       "| asv.29 | EE0493 |  2378 | EE0493_asv.29 |\n",
       "| asv.30 | EE0493 |     5 | EE0493_asv.30 |\n",
       "| asv.31 | EE0493 |  2645 | EE0493_asv.31 |\n",
       "| asv.33 | EE0493 |  2074 | EE0493_asv.33 |\n",
       "| asv.37 | EE0493 |  2200 | EE0493_asv.37 |\n",
       "| asv.38 | EE0493 |  1204 | EE0493_asv.38 |\n",
       "| asv.39 | EE0493 |    18 | EE0493_asv.39 |\n",
       "| asv.42 | EE0493 |     7 | EE0493_asv.42 |\n",
       "| asv.43 | EE0493 |   634 | EE0493_asv.43 |\n",
       "| asv.45 | EE0493 |    71 | EE0493_asv.45 |\n",
       "| asv.47 | EE0493 |   422 | EE0493_asv.47 |\n",
       "| asv.49 | EE0493 |     1 | EE0493_asv.49 |\n",
       "| asv.50 | EE0493 |     5 | EE0493_asv.50 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| asv.16798 | EE0495 | 11 | EE0495_asv.16798 |\n",
       "| asv.16813 | EE0495 |  1 | EE0495_asv.16813 |\n",
       "| asv.16814 | EE0495 |  1 | EE0495_asv.16814 |\n",
       "| asv.16815 | EE0495 |  4 | EE0495_asv.16815 |\n",
       "| asv.16816 | EE0495 |  7 | EE0495_asv.16816 |\n",
       "| asv.16823 | EE0495 |  3 | EE0495_asv.16823 |\n",
       "| asv.16826 | EE0495 |  9 | EE0495_asv.16826 |\n",
       "| asv.16847 | EE0495 |  3 | EE0495_asv.16847 |\n",
       "| asv.16864 | EE0495 |  1 | EE0495_asv.16864 |\n",
       "| asv.16867 | EE0495 |  5 | EE0495_asv.16867 |\n",
       "| asv.16875 | EE0495 |  1 | EE0495_asv.16875 |\n",
       "| asv.16879 | EE0495 |  2 | EE0495_asv.16879 |\n",
       "| asv.16880 | EE0495 |  5 | EE0495_asv.16880 |\n",
       "| asv.16881 | EE0495 |  5 | EE0495_asv.16881 |\n",
       "| asv.16882 | EE0495 |  3 | EE0495_asv.16882 |\n",
       "| asv.16897 | EE0495 |  2 | EE0495_asv.16897 |\n",
       "| asv.16898 | EE0495 |  1 | EE0495_asv.16898 |\n",
       "| asv.16914 | EE0495 |  2 | EE0495_asv.16914 |\n",
       "| asv.16916 | EE0495 |  2 | EE0495_asv.16916 |\n",
       "| asv.16918 | EE0495 |  3 | EE0495_asv.16918 |\n",
       "| asv.16919 | EE0495 |  3 | EE0495_asv.16919 |\n",
       "| asv.16920 | EE0495 |  2 | EE0495_asv.16920 |\n",
       "| asv.16921 | EE0495 |  3 | EE0495_asv.16921 |\n",
       "| asv.16922 | EE0495 |  3 | EE0495_asv.16922 |\n",
       "| asv.16923 | EE0495 |  2 | EE0495_asv.16923 |\n",
       "| asv.16949 | EE0495 |  1 | EE0495_asv.16949 |\n",
       "| asv.16958 | EE0495 |  1 | EE0495_asv.16958 |\n",
       "| asv.16961 | EE0495 |  1 | EE0495_asv.16961 |\n",
       "| asv.16962 | EE0495 |  1 | EE0495_asv.16962 |\n",
       "| asv.16963 | EE0495 |  2 | EE0495_asv.16963 |\n",
       "\n"
      ],
      "text/plain": [
       "      asv       eventID organismQuantity occurrenceID    \n",
       "1     asv.2     EE0493     14            EE0493_asv.2    \n",
       "2     asv.5     EE0493  40587            EE0493_asv.5    \n",
       "3     asv.6     EE0493      7            EE0493_asv.6    \n",
       "4     asv.7     EE0493  29367            EE0493_asv.7    \n",
       "5     asv.8     EE0493  72378            EE0493_asv.8    \n",
       "6     asv.9     EE0493  35970            EE0493_asv.9    \n",
       "7     asv.10    EE0493     10            EE0493_asv.10   \n",
       "8     asv.11    EE0493   4375            EE0493_asv.11   \n",
       "9     asv.12    EE0493   6853            EE0493_asv.12   \n",
       "10    asv.14    EE0493   2299            EE0493_asv.14   \n",
       "11    asv.16    EE0493     18            EE0493_asv.16   \n",
       "12    asv.17    EE0493   8017            EE0493_asv.17   \n",
       "13    asv.19    EE0493   1736            EE0493_asv.19   \n",
       "14    asv.20    EE0493    428            EE0493_asv.20   \n",
       "15    asv.21    EE0493   1711            EE0493_asv.21   \n",
       "16    asv.25    EE0493   2539            EE0493_asv.25   \n",
       "17    asv.28    EE0493      8            EE0493_asv.28   \n",
       "18    asv.29    EE0493   2378            EE0493_asv.29   \n",
       "19    asv.30    EE0493      5            EE0493_asv.30   \n",
       "20    asv.31    EE0493   2645            EE0493_asv.31   \n",
       "21    asv.33    EE0493   2074            EE0493_asv.33   \n",
       "22    asv.37    EE0493   2200            EE0493_asv.37   \n",
       "23    asv.38    EE0493   1204            EE0493_asv.38   \n",
       "24    asv.39    EE0493     18            EE0493_asv.39   \n",
       "25    asv.42    EE0493      7            EE0493_asv.42   \n",
       "26    asv.43    EE0493    634            EE0493_asv.43   \n",
       "27    asv.45    EE0493     71            EE0493_asv.45   \n",
       "28    asv.47    EE0493    422            EE0493_asv.47   \n",
       "29    asv.49    EE0493      1            EE0493_asv.49   \n",
       "30    asv.50    EE0493      5            EE0493_asv.50   \n",
       "⋮     ⋮         ⋮       ⋮                ⋮               \n",
       "13346 asv.16798 EE0495  11               EE0495_asv.16798\n",
       "13347 asv.16813 EE0495   1               EE0495_asv.16813\n",
       "13348 asv.16814 EE0495   1               EE0495_asv.16814\n",
       "13349 asv.16815 EE0495   4               EE0495_asv.16815\n",
       "13350 asv.16816 EE0495   7               EE0495_asv.16816\n",
       "13351 asv.16823 EE0495   3               EE0495_asv.16823\n",
       "13352 asv.16826 EE0495   9               EE0495_asv.16826\n",
       "13353 asv.16847 EE0495   3               EE0495_asv.16847\n",
       "13354 asv.16864 EE0495   1               EE0495_asv.16864\n",
       "13355 asv.16867 EE0495   5               EE0495_asv.16867\n",
       "13356 asv.16875 EE0495   1               EE0495_asv.16875\n",
       "13357 asv.16879 EE0495   2               EE0495_asv.16879\n",
       "13358 asv.16880 EE0495   5               EE0495_asv.16880\n",
       "13359 asv.16881 EE0495   5               EE0495_asv.16881\n",
       "13360 asv.16882 EE0495   3               EE0495_asv.16882\n",
       "13361 asv.16897 EE0495   2               EE0495_asv.16897\n",
       "13362 asv.16898 EE0495   1               EE0495_asv.16898\n",
       "13363 asv.16914 EE0495   2               EE0495_asv.16914\n",
       "13364 asv.16916 EE0495   2               EE0495_asv.16916\n",
       "13365 asv.16918 EE0495   3               EE0495_asv.16918\n",
       "13366 asv.16919 EE0495   3               EE0495_asv.16919\n",
       "13367 asv.16920 EE0495   2               EE0495_asv.16920\n",
       "13368 asv.16921 EE0495   3               EE0495_asv.16921\n",
       "13369 asv.16922 EE0495   3               EE0495_asv.16922\n",
       "13370 asv.16923 EE0495   2               EE0495_asv.16923\n",
       "13371 asv.16949 EE0495   1               EE0495_asv.16949\n",
       "13372 asv.16958 EE0495   1               EE0495_asv.16958\n",
       "13373 asv.16961 EE0495   1               EE0495_asv.16961\n",
       "13374 asv.16962 EE0495   1               EE0495_asv.16962\n",
       "13375 asv.16963 EE0495   2               EE0495_asv.16963"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "library(tidyr)\n",
    "\n",
    "seqtab %>%\n",
    "    gather(eventID, organismQuantity, 2:3) %>%\n",
    "    filter(organismQuantity > 0) %>%\n",
    "    mutate(occurrenceID = paste0(eventID, \"_\", asv))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9094e794-e8a7-471b-8b6c-baf7c06e4791",
   "metadata": {},
   "source": [
    "## Quality control\n",
    "\n",
    "### Taxon matching\n",
    "\n",
    "Let's first match the taxa with WoRMS. This can be done using the `obistools` package. Before matching with WoRMS we will remove underscores from the scientific names."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7c36c99f-04a1-480d-95d0-09527d254e4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "taxonomy <- taxonomy %>%\n",
    "    mutate(taxonomy = stringr::str_replace(taxonomy, \"_\", \" \"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "8f21e31c-9515-4646-b333-8c4c81da399f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "668 names, 0 without matches, 11 with multiple matches\n",
      "\n"
     ]
    }
   ],
   "source": [
    "matched <- obistools::match_taxa(taxonomy$taxonomy, ask = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "27c38f2f-1373-4641-b716-0afdae87e113",
   "metadata": {},
   "outputs": [],
   "source": [
    "taxonomy <- bind_cols(taxonomy, matched)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "2133f1b5-4a6a-47f8-a2a9-6973cb1bec08",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 36 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>taxonomy</th><th scope=col>n</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>undef Oomycota                </td><td>31</td></tr>\n",
       "\t<tr><td>Cypretta maya                 </td><td>21</td></tr>\n",
       "\t<tr><td>undef Bacteria_bacteria       </td><td>19</td></tr>\n",
       "\t<tr><td>Navicula                      </td><td> 8</td></tr>\n",
       "\t<tr><td>Clathria genus                </td><td> 6</td></tr>\n",
       "\t<tr><td>Aerococcus urinae             </td><td> 5</td></tr>\n",
       "\t<tr><td>Paravannella minima           </td><td> 4</td></tr>\n",
       "\t<tr><td>Calonectria colhounii         </td><td> 3</td></tr>\n",
       "\t<tr><td>Culex impudicus               </td><td> 3</td></tr>\n",
       "\t<tr><td>Nitzschia sp._BOLD:AAO7110    </td><td> 3</td></tr>\n",
       "\t<tr><td>undef Gastropoda              </td><td> 3</td></tr>\n",
       "\t<tr><td>Holothuria                    </td><td> 2</td></tr>\n",
       "\t<tr><td>Pinnularia acrosphaeria       </td><td> 2</td></tr>\n",
       "\t<tr><td>undef undef_Eukaryota         </td><td> 2</td></tr>\n",
       "\t<tr><td>Aplysia                       </td><td> 1</td></tr>\n",
       "\t<tr><td>Aurantiochytrium mangrovei    </td><td> 1</td></tr>\n",
       "\t<tr><td>Austrosciara hyalipennis      </td><td> 1</td></tr>\n",
       "\t<tr><td>Cerithium nodulosum           </td><td> 1</td></tr>\n",
       "\t<tr><td>Chondrilla sponges            </td><td> 1</td></tr>\n",
       "\t<tr><td>Cladonema digitatum           </td><td> 1</td></tr>\n",
       "\t<tr><td>Cladosporium tenuissimum      </td><td> 1</td></tr>\n",
       "\t<tr><td>Clytia                        </td><td> 1</td></tr>\n",
       "\t<tr><td>Cochliopodium larifeili       </td><td> 1</td></tr>\n",
       "\t<tr><td>Cochliopodium marrii          </td><td> 1</td></tr>\n",
       "\t<tr><td>Colletotrichum gloeosporioides</td><td> 1</td></tr>\n",
       "\t<tr><td>Cymadusa filosa               </td><td> 1</td></tr>\n",
       "\t<tr><td>Gerres longirostris           </td><td> 1</td></tr>\n",
       "\t<tr><td>Gliocladium viride            </td><td> 1</td></tr>\n",
       "\t<tr><td>Lobophora brown_algae         </td><td> 1</td></tr>\n",
       "\t<tr><td>Macroconia gigas              </td><td> 1</td></tr>\n",
       "\t<tr><td>Nitzschia diatoms             </td><td> 1</td></tr>\n",
       "\t<tr><td>Periplaneta brunnea           </td><td> 1</td></tr>\n",
       "\t<tr><td>Synchaetomella acerina        </td><td> 1</td></tr>\n",
       "\t<tr><td>Tapinoma melanocephalum       </td><td> 1</td></tr>\n",
       "\t<tr><td>Tetramorium scytalum          </td><td> 1</td></tr>\n",
       "\t<tr><td>Thoosa                        </td><td> 1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 36 × 2\n",
       "\\begin{tabular}{ll}\n",
       " taxonomy & n\\\\\n",
       " <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t undef Oomycota                 & 31\\\\\n",
       "\t Cypretta maya                  & 21\\\\\n",
       "\t undef Bacteria\\_bacteria        & 19\\\\\n",
       "\t Navicula                       &  8\\\\\n",
       "\t Clathria genus                 &  6\\\\\n",
       "\t Aerococcus urinae              &  5\\\\\n",
       "\t Paravannella minima            &  4\\\\\n",
       "\t Calonectria colhounii          &  3\\\\\n",
       "\t Culex impudicus                &  3\\\\\n",
       "\t Nitzschia sp.\\_BOLD:AAO7110     &  3\\\\\n",
       "\t undef Gastropoda               &  3\\\\\n",
       "\t Holothuria                     &  2\\\\\n",
       "\t Pinnularia acrosphaeria        &  2\\\\\n",
       "\t undef undef\\_Eukaryota          &  2\\\\\n",
       "\t Aplysia                        &  1\\\\\n",
       "\t Aurantiochytrium mangrovei     &  1\\\\\n",
       "\t Austrosciara hyalipennis       &  1\\\\\n",
       "\t Cerithium nodulosum            &  1\\\\\n",
       "\t Chondrilla sponges             &  1\\\\\n",
       "\t Cladonema digitatum            &  1\\\\\n",
       "\t Cladosporium tenuissimum       &  1\\\\\n",
       "\t Clytia                         &  1\\\\\n",
       "\t Cochliopodium larifeili        &  1\\\\\n",
       "\t Cochliopodium marrii           &  1\\\\\n",
       "\t Colletotrichum gloeosporioides &  1\\\\\n",
       "\t Cymadusa filosa                &  1\\\\\n",
       "\t Gerres longirostris            &  1\\\\\n",
       "\t Gliocladium viride             &  1\\\\\n",
       "\t Lobophora brown\\_algae          &  1\\\\\n",
       "\t Macroconia gigas               &  1\\\\\n",
       "\t Nitzschia diatoms              &  1\\\\\n",
       "\t Periplaneta brunnea            &  1\\\\\n",
       "\t Synchaetomella acerina         &  1\\\\\n",
       "\t Tapinoma melanocephalum        &  1\\\\\n",
       "\t Tetramorium scytalum           &  1\\\\\n",
       "\t Thoosa                         &  1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 36 × 2\n",
       "\n",
       "| taxonomy &lt;chr&gt; | n &lt;int&gt; |\n",
       "|---|---|\n",
       "| undef Oomycota                 | 31 |\n",
       "| Cypretta maya                  | 21 |\n",
       "| undef Bacteria_bacteria        | 19 |\n",
       "| Navicula                       |  8 |\n",
       "| Clathria genus                 |  6 |\n",
       "| Aerococcus urinae              |  5 |\n",
       "| Paravannella minima            |  4 |\n",
       "| Calonectria colhounii          |  3 |\n",
       "| Culex impudicus                |  3 |\n",
       "| Nitzschia sp._BOLD:AAO7110     |  3 |\n",
       "| undef Gastropoda               |  3 |\n",
       "| Holothuria                     |  2 |\n",
       "| Pinnularia acrosphaeria        |  2 |\n",
       "| undef undef_Eukaryota          |  2 |\n",
       "| Aplysia                        |  1 |\n",
       "| Aurantiochytrium mangrovei     |  1 |\n",
       "| Austrosciara hyalipennis       |  1 |\n",
       "| Cerithium nodulosum            |  1 |\n",
       "| Chondrilla sponges             |  1 |\n",
       "| Cladonema digitatum            |  1 |\n",
       "| Cladosporium tenuissimum       |  1 |\n",
       "| Clytia                         |  1 |\n",
       "| Cochliopodium larifeili        |  1 |\n",
       "| Cochliopodium marrii           |  1 |\n",
       "| Colletotrichum gloeosporioides |  1 |\n",
       "| Cymadusa filosa                |  1 |\n",
       "| Gerres longirostris            |  1 |\n",
       "| Gliocladium viride             |  1 |\n",
       "| Lobophora brown_algae          |  1 |\n",
       "| Macroconia gigas               |  1 |\n",
       "| Nitzschia diatoms              |  1 |\n",
       "| Periplaneta brunnea            |  1 |\n",
       "| Synchaetomella acerina         |  1 |\n",
       "| Tapinoma melanocephalum        |  1 |\n",
       "| Tetramorium scytalum           |  1 |\n",
       "| Thoosa                         |  1 |\n",
       "\n"
      ],
      "text/plain": [
       "   taxonomy                       n \n",
       "1  undef Oomycota                 31\n",
       "2  Cypretta maya                  21\n",
       "3  undef Bacteria_bacteria        19\n",
       "4  Navicula                        8\n",
       "5  Clathria genus                  6\n",
       "6  Aerococcus urinae               5\n",
       "7  Paravannella minima             4\n",
       "8  Calonectria colhounii           3\n",
       "9  Culex impudicus                 3\n",
       "10 Nitzschia sp._BOLD:AAO7110      3\n",
       "11 undef Gastropoda                3\n",
       "12 Holothuria                      2\n",
       "13 Pinnularia acrosphaeria         2\n",
       "14 undef undef_Eukaryota           2\n",
       "15 Aplysia                         1\n",
       "16 Aurantiochytrium mangrovei      1\n",
       "17 Austrosciara hyalipennis        1\n",
       "18 Cerithium nodulosum             1\n",
       "19 Chondrilla sponges              1\n",
       "20 Cladonema digitatum             1\n",
       "21 Cladosporium tenuissimum        1\n",
       "22 Clytia                          1\n",
       "23 Cochliopodium larifeili         1\n",
       "24 Cochliopodium marrii            1\n",
       "25 Colletotrichum gloeosporioides  1\n",
       "26 Cymadusa filosa                 1\n",
       "27 Gerres longirostris             1\n",
       "28 Gliocladium viride              1\n",
       "29 Lobophora brown_algae           1\n",
       "30 Macroconia gigas                1\n",
       "31 Nitzschia diatoms               1\n",
       "32 Periplaneta brunnea             1\n",
       "33 Synchaetomella acerina          1\n",
       "34 Tapinoma melanocephalum         1\n",
       "35 Tetramorium scytalum            1\n",
       "36 Thoosa                          1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "taxonomy %>%\n",
    "    filter(is.na(scientificNameID)) %>%\n",
    "    group_by(taxonomy) %>%\n",
    "    summarize(n = n()) %>%\n",
    "    arrange(desc(n))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8d36c10-345e-4e0b-8cb6-ea1c71b5fdee",
   "metadata": {},
   "source": [
    "Normally we have to resolve these names one by one, but for this exercise we will just fix the most common errors. For example, records annotated as eukaryotes can be populated with scientificName `Incertae sedis` and scientificNameID `urn:lsid:marinespecies.org:taxname:12`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "9745ca20-4d92-431f-9336-6068f38b17fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "taxonomy <- taxonomy %>%\n",
    "    mutate(\n",
    "        scientificName = case_when(taxonomy %in% c(\"Eukaryota\", \"undef Eukaryota\", \"\") ~ \"Incertae sedis\", .default = scientificName),\n",
    "        scientificNameID = case_when(taxonomy %in% c(\"Eukaryota\", \"undef Eukaryota\", \"\") ~ \"urn:lsid:marinespecies.org:taxname:12\", .default = scientificNameID)\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "1186640b-112f-4dbe-848c-4095b07c6c50",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 36 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>taxonomy</th><th scope=col>n</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>undef Oomycota                </td><td>31</td></tr>\n",
       "\t<tr><td>Cypretta maya                 </td><td>21</td></tr>\n",
       "\t<tr><td>undef Bacteria_bacteria       </td><td>19</td></tr>\n",
       "\t<tr><td>Navicula                      </td><td> 8</td></tr>\n",
       "\t<tr><td>Clathria genus                </td><td> 6</td></tr>\n",
       "\t<tr><td>Aerococcus urinae             </td><td> 5</td></tr>\n",
       "\t<tr><td>Paravannella minima           </td><td> 4</td></tr>\n",
       "\t<tr><td>Calonectria colhounii         </td><td> 3</td></tr>\n",
       "\t<tr><td>Culex impudicus               </td><td> 3</td></tr>\n",
       "\t<tr><td>Nitzschia sp._BOLD:AAO7110    </td><td> 3</td></tr>\n",
       "\t<tr><td>undef Gastropoda              </td><td> 3</td></tr>\n",
       "\t<tr><td>Holothuria                    </td><td> 2</td></tr>\n",
       "\t<tr><td>Pinnularia acrosphaeria       </td><td> 2</td></tr>\n",
       "\t<tr><td>undef undef_Eukaryota         </td><td> 2</td></tr>\n",
       "\t<tr><td>Aplysia                       </td><td> 1</td></tr>\n",
       "\t<tr><td>Aurantiochytrium mangrovei    </td><td> 1</td></tr>\n",
       "\t<tr><td>Austrosciara hyalipennis      </td><td> 1</td></tr>\n",
       "\t<tr><td>Cerithium nodulosum           </td><td> 1</td></tr>\n",
       "\t<tr><td>Chondrilla sponges            </td><td> 1</td></tr>\n",
       "\t<tr><td>Cladonema digitatum           </td><td> 1</td></tr>\n",
       "\t<tr><td>Cladosporium tenuissimum      </td><td> 1</td></tr>\n",
       "\t<tr><td>Clytia                        </td><td> 1</td></tr>\n",
       "\t<tr><td>Cochliopodium larifeili       </td><td> 1</td></tr>\n",
       "\t<tr><td>Cochliopodium marrii          </td><td> 1</td></tr>\n",
       "\t<tr><td>Colletotrichum gloeosporioides</td><td> 1</td></tr>\n",
       "\t<tr><td>Cymadusa filosa               </td><td> 1</td></tr>\n",
       "\t<tr><td>Gerres longirostris           </td><td> 1</td></tr>\n",
       "\t<tr><td>Gliocladium viride            </td><td> 1</td></tr>\n",
       "\t<tr><td>Lobophora brown_algae         </td><td> 1</td></tr>\n",
       "\t<tr><td>Macroconia gigas              </td><td> 1</td></tr>\n",
       "\t<tr><td>Nitzschia diatoms             </td><td> 1</td></tr>\n",
       "\t<tr><td>Periplaneta brunnea           </td><td> 1</td></tr>\n",
       "\t<tr><td>Synchaetomella acerina        </td><td> 1</td></tr>\n",
       "\t<tr><td>Tapinoma melanocephalum       </td><td> 1</td></tr>\n",
       "\t<tr><td>Tetramorium scytalum          </td><td> 1</td></tr>\n",
       "\t<tr><td>Thoosa                        </td><td> 1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 36 × 2\n",
       "\\begin{tabular}{ll}\n",
       " taxonomy & n\\\\\n",
       " <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t undef Oomycota                 & 31\\\\\n",
       "\t Cypretta maya                  & 21\\\\\n",
       "\t undef Bacteria\\_bacteria        & 19\\\\\n",
       "\t Navicula                       &  8\\\\\n",
       "\t Clathria genus                 &  6\\\\\n",
       "\t Aerococcus urinae              &  5\\\\\n",
       "\t Paravannella minima            &  4\\\\\n",
       "\t Calonectria colhounii          &  3\\\\\n",
       "\t Culex impudicus                &  3\\\\\n",
       "\t Nitzschia sp.\\_BOLD:AAO7110     &  3\\\\\n",
       "\t undef Gastropoda               &  3\\\\\n",
       "\t Holothuria                     &  2\\\\\n",
       "\t Pinnularia acrosphaeria        &  2\\\\\n",
       "\t undef undef\\_Eukaryota          &  2\\\\\n",
       "\t Aplysia                        &  1\\\\\n",
       "\t Aurantiochytrium mangrovei     &  1\\\\\n",
       "\t Austrosciara hyalipennis       &  1\\\\\n",
       "\t Cerithium nodulosum            &  1\\\\\n",
       "\t Chondrilla sponges             &  1\\\\\n",
       "\t Cladonema digitatum            &  1\\\\\n",
       "\t Cladosporium tenuissimum       &  1\\\\\n",
       "\t Clytia                         &  1\\\\\n",
       "\t Cochliopodium larifeili        &  1\\\\\n",
       "\t Cochliopodium marrii           &  1\\\\\n",
       "\t Colletotrichum gloeosporioides &  1\\\\\n",
       "\t Cymadusa filosa                &  1\\\\\n",
       "\t Gerres longirostris            &  1\\\\\n",
       "\t Gliocladium viride             &  1\\\\\n",
       "\t Lobophora brown\\_algae          &  1\\\\\n",
       "\t Macroconia gigas               &  1\\\\\n",
       "\t Nitzschia diatoms              &  1\\\\\n",
       "\t Periplaneta brunnea            &  1\\\\\n",
       "\t Synchaetomella acerina         &  1\\\\\n",
       "\t Tapinoma melanocephalum        &  1\\\\\n",
       "\t Tetramorium scytalum           &  1\\\\\n",
       "\t Thoosa                         &  1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 36 × 2\n",
       "\n",
       "| taxonomy &lt;chr&gt; | n &lt;int&gt; |\n",
       "|---|---|\n",
       "| undef Oomycota                 | 31 |\n",
       "| Cypretta maya                  | 21 |\n",
       "| undef Bacteria_bacteria        | 19 |\n",
       "| Navicula                       |  8 |\n",
       "| Clathria genus                 |  6 |\n",
       "| Aerococcus urinae              |  5 |\n",
       "| Paravannella minima            |  4 |\n",
       "| Calonectria colhounii          |  3 |\n",
       "| Culex impudicus                |  3 |\n",
       "| Nitzschia sp._BOLD:AAO7110     |  3 |\n",
       "| undef Gastropoda               |  3 |\n",
       "| Holothuria                     |  2 |\n",
       "| Pinnularia acrosphaeria        |  2 |\n",
       "| undef undef_Eukaryota          |  2 |\n",
       "| Aplysia                        |  1 |\n",
       "| Aurantiochytrium mangrovei     |  1 |\n",
       "| Austrosciara hyalipennis       |  1 |\n",
       "| Cerithium nodulosum            |  1 |\n",
       "| Chondrilla sponges             |  1 |\n",
       "| Cladonema digitatum            |  1 |\n",
       "| Cladosporium tenuissimum       |  1 |\n",
       "| Clytia                         |  1 |\n",
       "| Cochliopodium larifeili        |  1 |\n",
       "| Cochliopodium marrii           |  1 |\n",
       "| Colletotrichum gloeosporioides |  1 |\n",
       "| Cymadusa filosa                |  1 |\n",
       "| Gerres longirostris            |  1 |\n",
       "| Gliocladium viride             |  1 |\n",
       "| Lobophora brown_algae          |  1 |\n",
       "| Macroconia gigas               |  1 |\n",
       "| Nitzschia diatoms              |  1 |\n",
       "| Periplaneta brunnea            |  1 |\n",
       "| Synchaetomella acerina         |  1 |\n",
       "| Tapinoma melanocephalum        |  1 |\n",
       "| Tetramorium scytalum           |  1 |\n",
       "| Thoosa                         |  1 |\n",
       "\n"
      ],
      "text/plain": [
       "   taxonomy                       n \n",
       "1  undef Oomycota                 31\n",
       "2  Cypretta maya                  21\n",
       "3  undef Bacteria_bacteria        19\n",
       "4  Navicula                        8\n",
       "5  Clathria genus                  6\n",
       "6  Aerococcus urinae               5\n",
       "7  Paravannella minima             4\n",
       "8  Calonectria colhounii           3\n",
       "9  Culex impudicus                 3\n",
       "10 Nitzschia sp._BOLD:AAO7110      3\n",
       "11 undef Gastropoda                3\n",
       "12 Holothuria                      2\n",
       "13 Pinnularia acrosphaeria         2\n",
       "14 undef undef_Eukaryota           2\n",
       "15 Aplysia                         1\n",
       "16 Aurantiochytrium mangrovei      1\n",
       "17 Austrosciara hyalipennis        1\n",
       "18 Cerithium nodulosum             1\n",
       "19 Chondrilla sponges              1\n",
       "20 Cladonema digitatum             1\n",
       "21 Cladosporium tenuissimum        1\n",
       "22 Clytia                          1\n",
       "23 Cochliopodium larifeili         1\n",
       "24 Cochliopodium marrii            1\n",
       "25 Colletotrichum gloeosporioides  1\n",
       "26 Cymadusa filosa                 1\n",
       "27 Gerres longirostris             1\n",
       "28 Gliocladium viride              1\n",
       "29 Lobophora brown_algae           1\n",
       "30 Macroconia gigas                1\n",
       "31 Nitzschia diatoms               1\n",
       "32 Periplaneta brunnea             1\n",
       "33 Synchaetomella acerina          1\n",
       "34 Tapinoma melanocephalum         1\n",
       "35 Tetramorium scytalum            1\n",
       "36 Thoosa                          1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "taxonomy %>%\n",
    "    filter(is.na(scientificNameID)) %>%\n",
    "    group_by(taxonomy) %>%\n",
    "    summarize(n = n()) %>%\n",
    "    arrange(desc(n))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "45b978c2-0300-4cfc-8662-13ef85ac45d1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 6</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>asv</th><th scope=col>taxonomy</th><th scope=col>scientificName</th><th scope=col>scientificNameID</th><th scope=col>match_type</th><th scope=col>acceptedNameUsageID</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>248</th><td>asv.1</td><td>Eukaryota             </td><td>Incertae sedis        </td><td>urn:lsid:marinespecies.org:taxname:12    </td><td>NA   </td><td>NA    </td></tr>\n",
       "\t<tr><th scope=row>162</th><td>asv.2</td><td>Clausocalanus furcatus</td><td>Clausocalanus furcatus</td><td>urn:lsid:marinespecies.org:taxname:104503</td><td>exact</td><td>104503</td></tr>\n",
       "\t<tr><th scope=row>252</th><td>asv.3</td><td>Eurotatoria           </td><td>Eurotatoria           </td><td>urn:lsid:marinespecies.org:taxname:368537</td><td>exact</td><td>368537</td></tr>\n",
       "\t<tr><th scope=row>49</th><td>asv.4</td><td>Arthropoda            </td><td>Arthropoda            </td><td>urn:lsid:marinespecies.org:taxname:1065  </td><td>exact</td><td>1065  </td></tr>\n",
       "\t<tr><th scope=row>248.1</th><td>asv.5</td><td>Eukaryota             </td><td>Incertae sedis        </td><td>urn:lsid:marinespecies.org:taxname:12    </td><td>NA   </td><td>NA    </td></tr>\n",
       "\t<tr><th scope=row>255</th><td>asv.6</td><td>Farranula gibbula     </td><td>Farranula gibbula     </td><td>urn:lsid:marinespecies.org:taxname:346477</td><td>exact</td><td>346477</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 6\n",
       "\\begin{tabular}{r|llllll}\n",
       "  & asv & taxonomy & scientificName & scientificNameID & match\\_type & acceptedNameUsageID\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t248 & asv.1 & Eukaryota              & Incertae sedis         & urn:lsid:marinespecies.org:taxname:12     & NA    & NA    \\\\\n",
       "\t162 & asv.2 & Clausocalanus furcatus & Clausocalanus furcatus & urn:lsid:marinespecies.org:taxname:104503 & exact & 104503\\\\\n",
       "\t252 & asv.3 & Eurotatoria            & Eurotatoria            & urn:lsid:marinespecies.org:taxname:368537 & exact & 368537\\\\\n",
       "\t49 & asv.4 & Arthropoda             & Arthropoda             & urn:lsid:marinespecies.org:taxname:1065   & exact & 1065  \\\\\n",
       "\t248.1 & asv.5 & Eukaryota              & Incertae sedis         & urn:lsid:marinespecies.org:taxname:12     & NA    & NA    \\\\\n",
       "\t255 & asv.6 & Farranula gibbula      & Farranula gibbula      & urn:lsid:marinespecies.org:taxname:346477 & exact & 346477\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 6\n",
       "\n",
       "| <!--/--> | asv &lt;chr&gt; | taxonomy &lt;chr&gt; | scientificName &lt;chr&gt; | scientificNameID &lt;chr&gt; | match_type &lt;chr&gt; | acceptedNameUsageID &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|\n",
       "| 248 | asv.1 | Eukaryota              | Incertae sedis         | urn:lsid:marinespecies.org:taxname:12     | NA    | NA     |\n",
       "| 162 | asv.2 | Clausocalanus furcatus | Clausocalanus furcatus | urn:lsid:marinespecies.org:taxname:104503 | exact | 104503 |\n",
       "| 252 | asv.3 | Eurotatoria            | Eurotatoria            | urn:lsid:marinespecies.org:taxname:368537 | exact | 368537 |\n",
       "| 49 | asv.4 | Arthropoda             | Arthropoda             | urn:lsid:marinespecies.org:taxname:1065   | exact | 1065   |\n",
       "| 248.1 | asv.5 | Eukaryota              | Incertae sedis         | urn:lsid:marinespecies.org:taxname:12     | NA    | NA     |\n",
       "| 255 | asv.6 | Farranula gibbula      | Farranula gibbula      | urn:lsid:marinespecies.org:taxname:346477 | exact | 346477 |\n",
       "\n"
      ],
      "text/plain": [
       "      asv   taxonomy               scientificName        \n",
       "248   asv.1 Eukaryota              Incertae sedis        \n",
       "162   asv.2 Clausocalanus furcatus Clausocalanus furcatus\n",
       "252   asv.3 Eurotatoria            Eurotatoria           \n",
       "49    asv.4 Arthropoda             Arthropoda            \n",
       "248.1 asv.5 Eukaryota              Incertae sedis        \n",
       "255   asv.6 Farranula gibbula      Farranula gibbula     \n",
       "      scientificNameID                          match_type acceptedNameUsageID\n",
       "248   urn:lsid:marinespecies.org:taxname:12     NA         NA                 \n",
       "162   urn:lsid:marinespecies.org:taxname:104503 exact      104503             \n",
       "252   urn:lsid:marinespecies.org:taxname:368537 exact      368537             \n",
       "49    urn:lsid:marinespecies.org:taxname:1065   exact      1065               \n",
       "248.1 urn:lsid:marinespecies.org:taxname:12     NA         NA                 \n",
       "255   urn:lsid:marinespecies.org:taxname:346477 exact      346477             "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(taxonomy)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f8d97af-767e-4dee-ad19-b2a4bb205c39",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
